src/game/ZoneManager.cxx

```diff
@@ -0,0 +1,77 @@
+#include "ZoneManager.hxx"
+#include "Engine.hxx"
+#include "LOG.hxx"
+
+#include <algorithm>
+#include <random>
+
+void ZoneManager::updateZones()
+{
+  m_MapNodes.clear();
+  for (const MapNode &node : Engine::instance().map->getMapNodes())
+  {
+    if (node.getTileData(ZONE)) // if there's a zone this layer is not null
+    {
+      m_MapNodes.emplace_back(&node);
+    }
+  }
+}
+
+void ZoneManager::spawnBuildings()
+{
+  int amountOfBuildingsToSpawn = 5;
+  int buildingsSpawned = 0;
+
+  // shuffle mapNodes
+  auto rd = std::random_device{};
+  auto rng = std::default_random_engine{rd()};
+  std::shuffle(std::begin(m_MapNodes), std::end(m_MapNodes), rng);
+
+  // pick every single zone node we have
+  for (auto node : m_MapNodes)
+  {
+    std::vector<const std::string> myVec;
+    if (buildingsSpawned >= amountOfBuildingsToSpawn)
+    {
+      break;
+    }
+
+    if (!node->getTileData(Layer::ZONE))
+    {
+      LOG(LOG_ERROR) << "Something is wrong with a zone tile - Report this to the Cytopia Team!";
+      continue;
+    }
+    // get the right zone enum of the zone zile
+    for (auto zone : node->getTileData(Layer::ZONE)->zones)
+    {
+      if (TileManager::instance().getTileIDsOfCategory(zone).empty())
+      {
+        LOG(LOG_ERROR) << "No buildings available for zone: " << zone;
+        continue;
+      }
+      else
+      {
+        // add all the tileIDs that are available for this zone to our vector
+        for (auto ID : TileManager::instance().getTileIDsOfCategory(zone))
+        {
+          myVec.emplace_back(ID);
+        }
+      }
+    }
+
+    if (myVec.empty())
+    {
+      LOG(LOG_ERROR) << "No buildings available! Aborting!";
+      break;
+    }
+    // pick a random tileID from the vector
+    // TODO: Check how many neighboring tiles are available and pick a building with correct tilesize
+    int randomIndex = rand() % myVec.size();
+
+    std::string tileToPlace = myVec[randomIndex];
+    std::vector targetObjectNodes = Engine::instance().map->getObjectCoords(node->getCoordinates(), tileToPlace);
+    Engine::instance().setTileIDOfNode(targetObjectNodes.begin(), targetObjectNodes.end(), tileToPlace, false);
```

Comment:
we should start writing more clear code. tileToPlace -> building
The next two lines should be one function like placeBuilding(coordinates, building)

Reply:
good suggestion,
but I suggest even place or emplace, just to make it general, not only for buildings

Reply:
As a general idea, we need more code layers to hide this kind of complexity in gameplay (higher level) code. So this is a great place to start.

Reply:
renamed **tileToPlace** to **building**

Reply:
What about exacting the function?

Reply:
I don't understand what you mean? Please be more specific

Reply:
Lines 72 and 73 should be a function for itself. I can imagine the same type of code we have all around.